packages/connect-node/src/private/node-error.ts

```diff
@@ -25,14 +25,29 @@ import {
  */
 export function connectErrorFromNodeReason(reason: unknown): ConnectError {
   let code = Code.Internal;
-  const dnsResolveFailed = unwrapNodeErrorChain(reason)
-    .map((e) => getNodeErrorProps(e))
-    .some(
+  const chain = unwrapNodeErrorChain(reason).map(getNodeErrorProps);
+  if (chain.some((p) => p.code == "ERR_STREAM_WRITE_AFTER_END")) {
+    // We do not want intentional errors from the server to be shadowed
+    // by client-side errors.
+    // This can occur if the server has written a response with an error
+    // and has ended the connection. This response may already sit in a
+    // buffer on the client, while it is still writing to the request
+    // body.
+    // To avoid this problem, we wrap ERR_STREAM_WRITE_AFTER_END as a
+    // ConnectError with Code.Aborted. The special meaning of this code
+    // in this situation is documented in StreamingConn.send() and in
+    // createServerStreamingFn().
+    code = Code.Aborted;
```

Comment:
We convert ERR_STREAM_WRITE_AFTER_END to a Code.Aborted.

All transports were updated to wrap errors from using this function when writing to the request body.

Promise clients and callback clients special case Code.Aborted in their implementation for server-streaming methods so that `for await (const res of client.foo({})) {...}` has reasonable behavior.

promise-client.spec.ts of connect-core covers the client behavior. The stub transport introduced for that is now also used to add some test coverage for interceptors in connect-core.